File: src/main/java/org/example/bankramenserver/domain/recurring/domain/repository/RecurringPaymentRepository.java
package org.example.bankramenserver.domain.recurring.domain.repository;

import org.example.bankramenserver.domain.recurring.domain.RecurringPayment;
import org.springframework.data.jpa.repository.EntityGraph;
import org.springframework.data.jpa.repository.JpaRepository;

import java.time.LocalDateTime;
import java.util.List;
import java.util.UUID;

public interface RecurringPaymentRepository extends JpaRepository<RecurringPayment, UUID> {

    @EntityGraph(attributePaths = {"transactions", "transactions.transaction"})
    List<RecurringPayment> findAllByUser_IdAndActiveTrueOrderByNextBillingDateAsc(UUID userId);

    List<RecurringPayment> findAllByActiveTrueAndConfirmedTrueAndNextBillingDateBetween(
            LocalDateTime start,
            LocalDateTime end
    );

    List<RecurringPayment> findAllByUser_IdAndActiveTrueAndConfirmedTrueAndNextBillingDateBetween(
            UUID userId,
            LocalDateTime start,
            LocalDateTime end
    );

    boolean existsByUser_IdAndNameAndAmountAndCycleAndActiveTrue(
            UUID userId,
            String name,
            Long amount,
            RecurringPayment.Cycle cycle
    );

    List<RecurringPayment> findAllByActiveTrueAndConfirmedTrueAndNextBillingDateGreaterThanEqualAndNextBillingDateLessThan(
            LocalDateTime start,
            LocalDateTime end
    );
}

File: src/main/kotlin/org/example/balogserver/domain/recurring/domain/repository/RecurringPaymentRepository.kt
package org.example.balogserver.domain.recurring.domain.repository

import org.example.balogserver.domain.recurring.domain.RecurringPayment
import org.springframework.data.jpa.repository.EntityGraph
import org.springframework.data.jpa.repository.JpaRepository
import java.time.LocalDateTime
import java.util.UUID

interface RecurringPaymentRepository : JpaRepository<RecurringPayment, UUID> {
    @EntityGraph(attributePaths = ["transactions", "transactions.transaction"])
    fun findAllByUser_IdAndActiveTrueOrderByNextBillingDateAsc(userId: UUID?): List<RecurringPayment>
    fun findAllByActiveTrueAndConfirmedTrueAndNextBillingDateBetween(start: LocalDateTime?, end: LocalDateTime?): List<RecurringPayment>
    fun findAllByUser_IdAndActiveTrueAndConfirmedTrueAndNextBillingDateBetween(userId: UUID?, start: LocalDateTime?, end: LocalDateTime?): List<RecurringPayment>
    fun existsByUser_IdAndNameAndAmountAndCycleAndActiveTrue(userId: UUID?, name: String?, amount: Long?, cycle: RecurringPayment.Cycle?): Boolean
    fun findAllByActiveTrueAndConfirmedTrueAndNextBillingDateGreaterThanEqualAndNextBillingDateLessThan(start: LocalDateTime?, end: LocalDateTime?): List<RecurringPayment>
}
